drivers/adc/adc_ad4114.c

```diff
@@ -0,0 +1,462 @@
+{
+	struct adc_ad4114_data *data = dev->data;
+	const struct adc_ad4114_config *config = dev->config;
+	int ret;
+	uint8_t write_reg[2];
+	uint8_t status;
+
+	ret = adc_ad4114x_validate_buffer_size(dev, sequence);
+	if (ret < 0) {
+		LOG_ERR("insufficient buffer size");
+		return ret;
+	}
+
+	data->channels_cfg = sequence->channels;
+	for (uint32_t i = 0U; i < AD4114_CHAN_NUMBER; i++) {
+		if ((BIT(i) & sequence->channels) != 0) {
+			write_reg[0] = 0x80 | (uint8_t)((config->map_input[i] >> 8) & 0xFF);
+			write_reg[1] = (uint8_t)(config->map_input[i] & 0xFF);
+			LOG_DBG("Enable channel %d with mapping %X %X, raw %X", i, write_reg[0],
+				write_reg[1], config->map_input[i]);
+			ad4114_write_reg(dev, AD4114_CHANNEL_0_REG + i, write_reg, 2);
+		} else {
+			LOG_DBG("Disable channel %d", i);
+			write_reg[0] = 0x0;
+			write_reg[1] = 0x0;
+			ad4114_write_reg(dev, AD4114_CHANNEL_0_REG + i, write_reg, 2);
+		}
+	}
+
+	/* Configure the buffer */
+	data->buffer = sequence->buffer;
+
+	while ((status & 0x80) != 0x80) {
+		/* Wait for acquiistion start */
+		ad4114_read_reg(dev, AD4114_STATUS_REG, &status, 1);
+		k_usleep(10);
+	}
+
+	adc_context_start_read(&data->ctx, sequence);
+
+	return adc_context_wait_for_completion(&data->ctx);
+}
+
+static void adc_ad4114_acquisition_thread(struct adc_ad4114_data *data)
+{
+	uint8_t value[4] = {0};
+	uint32_t buffer_values[AD4114_CHAN_NUMBER];
+	bool is_ended = false;
+
+	while (true) {
+		k_sem_take(&data->sem, K_FOREVER);
+
+		while (data->channels != 0) {
+			ad4114_read_reg(data->dev, AD4114_DATA_REG, value, 4);
+			/* Check the read channel */
+			if ((value[3] & 0xF0) != 0) {
+				LOG_DBG("Error read on :  %X ", value[3]);
+			} else {
+				LOG_DBG("Success read on %d: value  %X ", value[3],
+					(value[2] << 16 | value[1] << 8 | value[0]));
+				/* success read, store it */
+				buffer_values[value[3]] =
+					(value[0] << 16 | value[1] << 8 | value[2]);
+				WRITE_BIT(data->channels, value[3], 0);
+				/* Disable the channel after read success */
+				uint8_t write_reg[2] = {0};
+
+				ad4114_write_reg(data->dev, AD4114_CHANNEL_0_REG + value[3],
+						 write_reg, 2);
+			}
+			if (data->channels == 0) {
+				is_ended = true;
+			}
+			k_usleep(10);
+		}
+
+		if (is_ended) {
+			is_ended = false;
+			for (uint8_t i = 0U; i < AD4114_CHAN_NUMBER; i++) {
+				if ((BIT(i) & data->channels_cfg) != 0) {
+					*data->buffer++ = buffer_values[i];
+					LOG_DBG("Read channel %d value :  %X ", i,
+						buffer_values[i]);
+				}
+			}
+			adc_context_on_sampling_done(&data->ctx, data->dev);
+		}
+		k_usleep(1000);
+	}
+}
+
+static int adc_ad4114_channel_setup(const struct device *dev,
+				    const struct adc_channel_cfg *channel_cfg)
+{
+
+	/* Todo in the futur we can manage here :
+	 * filters
+	 * gain
+	 * offsets
+	 * special configuration : we can update map_input here to override the device
+	 * tree setup
+	 */
```

Comment:
Return an errno for channel configurations that aren't supported

Reply:
done